Add initial NUMA awareness support

debrouxl opened the pull request:
Here's a cleaned up version of my NUMA awareness work, where some changes such as unconditional traces or nops to help setting debugger breakpoints have been removed, and other changes meant to help debug and validation are still available for now, but commented out. The commented out benchmarking hunk from app/display.c and the ACPI SLIT handling are there as well; at least the former needs to stay, I use it constantly.
I decoupled this commit from other changes (nontemporal, partial NX support, an optimization); the combination remains available on the numa_combined branch, also containing unconditional traces and nops.

See #12 for performance measurements on a number of computers I have access to, if you want to read them again. On 2S and 4S servers, it almost always yields a wall clock time reduction (no recent test on the 2S Westmere), occasionally an very high one on a ring topology server. Memory is therefore tested at a lower latency and higher bandwidth, because reads and writes are not bottlenecked by inter-socket and intra-socket links. Also, with in the default powersave=high mode, power consumption is reduced.

A couple implementation notes:
* in the absence of a SRAT in the ACPI data, smp_init() contains a small bit of code to emulate a single-node NUMA setup. I didn't set enable_numa to false in this case, in order to avoid a reference from system to app;
* proximity domain handling costs 20 * MAX_APIC_IDS + 6 * MAX_PROXIMITY_DOMAINS BSS bytes. Having MAX_PROXIMITY_DOMAINS = MAX_APIC_IDS (system/smp.h) and MAX_APIC_IDs memory affinity ranges is certainly excessive, but who knows how many proximity domains and memory affinity ranges SRATs from 16S and 32S SPR-generation monster servers from HPE or Bull expose, even without CXL memory (which is partially defined in another table, IIUC)... Not that memtest86+ would run adequately on such servers as long as x2APIC mode is not supported anyway, but still.

NB: I haven't been able to test changes in error detection reliability or time to first error. Faulty sticks of the "desktop computer will emit loud beeps and refuse to boot" or "server will mask out the faulty stick and complain" varieties are useless for that purpose...

Fixes #12 .

martinwhitaker commented on system/smp.c:
```diff
@@ -548,13 +824,35 @@ void smp_init(bool smp_enable)
+
+    // Create dummy entries when there's no (proper) memory affinity information.
+    if (num_memory_affinity_ranges == 0 || num_proximity_domains == 0) {
+        extern uintptr_t pm_limit_lower;
```
Not convinced this is preferable to having a reference to `enable_numa` as you are still referencing variables in app!

I would have just set `num_proximity domains` to 0 and checked for that in `main()` after calling `smp_init()`.

debrouxl replied:
Heh, right. Missed that extern reference when writing the cover message. Will fix.

debrouxl replied:
Fixed by removing that block and setting `enable_numa` to `false` in `main()` below the call to `smp_init()`, which is itself below the call to `config_init()`.

martinwhitaker commented on app/main.c:
```diff
@@ -267,7 +269,13 @@ static void global_init(void)
-            chunk_index[i] = num_enabled_cpus;
+            if (enable_numa) {
+                uint32_t proximity_domain_idx = smp_get_proximity_domain_idx(i);
+                chunk_index_in_proximity_domain[i] = smp_alloc_cpu_in_proximity_domain(proximity_domain_idx);
```
Couldn't you have just used `chunk_index` for this case too?

debrouxl replied:
In the end, yes, it seems so. At first, I wasn't sure it would be possible not to use both `chunk_index` and `chunk_index_in_proximity_domain` in the same code path.

debrouxl replied:
Got rid of `chunk_index_in_proximity_domain`.

martinwhitaker commented on app/main.c:
```diff
@@ -369,13 +381,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
+                uintptr_t new_end;
+
+                while (1) {
+                    if (smp_narrow_to_proximity_domain(orig_start, orig_end, &proximity_domain_idx, &new_start, &new_end)) {
```
How often does a segment straddle proximity domains (given that the segment size is limited by the 1GB window)? If that is a rare case, this could be made much simpler by just allowing all CPUs whose proximity regions intersect the segment to work on that segment.

debrouxl replied:
All I know is that I tried to play safe, and that I'm pretty sure I debugged through that code path on QEMU :)
The prime contributor to 1 GB segments straddling across proximity domains is probably the memory hole.

martinwhitaker reviewed (commented):
Several substantive comments plus some whingeing about coding style :-)

debrouxl commented on app/main.c:
```diff
@@ -267,7 +269,13 @@ static void global_init(void)
+                uint32_t proximity_domain_idx = smp_get_proximity_domain_idx(i);
+                chunk_index_in_proximity_domain[i] = smp_alloc_cpu_in_proximity_domain(proximity_domain_idx);
+            }
+            else {
```
Crap. I keep forgetting about that one. It's basically the only difference between my usual coding style and the memtest86+ coding style :)

debrouxl commented on app/main.c:
```diff
@@ -369,13 +381,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
+            }
+            else {
+                // Now also pay attention to proximity domains, which are based on physical addresses.
+                uintptr_t orig_start = seg_start << PAGE_SHIFT;
```
Good question. I guess I have to check whether the ACPI spec mandates that proximity region addresses are page-aligned... but BIOS manufacturers do not always follow specs, anyway.

debrouxl replied:
I don't see anything like that in section 5.2.16 of [the ACPI SPEC](https://uefi.org/htmlspecs/ACPI_Spec_6_4_html/05_ACPI_Software_Programming_Model/ACPI_Software_Programming_Model.html), unfortunately.

debrouxl replied:
Switched to `uint64_t`.

debrouxl commented on system/smp.c:
```diff
@@ -427,6 +503,187 @@ static bool find_cpus_in_madt(void)
+                }
+                if (found == -1) {
+                    // Not found, allocate entry.
+                    if (num_proximity_domains < (int)(sizeof(proximity_domains) / sizeof(proximity_domains[0]))) {
```
Right, a macro named e.g. ARRAY_SIZE, as in some other code bases, would be better.

debrouxl replied:
I created an ARRAY_SIZE macro. `boot/boot.h` might be slightly too high in the includes hierarchy for that, though. Once the location is settled, I can split its addition to another, earlier commit and use it in `system/smbus.c` for `intel_ich5_dids`.

martinwhitaker replied:
I think it should go in the lib directory, but there isn't an existing header file that seems appropriate. Maybe create a `macros.h` so we can add other things as and when they occur to us.

debrouxl replied:
I did that, and switched to the safer, if less portable (GNU extension), definition for `ARRAY_SIZE()` found in the Linux kernel.

debrouxl commented on system/smp.c:
```diff
@@ -427,6 +503,187 @@ static bool find_cpus_in_madt(void)
+            srat_memory_affinity_entry_t *entry = (srat_memory_affinity_entry_t *)tab_entry_ptr;
+            if (entry->flags & SRAT_MAF_ENABLED) {
+                uint32_t proximity_domain = entry->proximity_domain;
+                uintptr_t start = entry->base_address;
```
Oh, yeah. This is ample proof to the fact that I did not give much thought about NUMA on 32-bit x86, as the Opterons which introduced NUMA were the first x86-64 platforms, and on the Intel side, Nehalem are x86-64...
_If_ all known x86 NUMA platforms are x86-64, then it is technically _possible_ to make this NUMA awareness code x86-64-only... but, er, I'm not going to dare pretend that sprinking `#ifdef __x86_64__` throughout the code, or probably worse, exporting some of the relevant bits of code to static inline functions defined a header file just in order to have fewer such preprocessor directives, is _desirable_ :)

debrouxl replied:
Switched to uint64_t.

debrouxl commented on app/main.c:
```diff
@@ -369,13 +381,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
+            }
         }
     }
+    /*for (int i = 0; i < vm_map_size; i++) {
```
Right, I did that in another hunk but failed to do it here.

debrouxl commented on app/main.c:
```diff
@@ -113,6 +114,7 @@ barrier_t   *run_barrier = NULL;
 spinlock_t  *error_mutex = NULL;
 
 vm_map_t    vm_map[MAX_MEM_SEGMENTS];
+uint32_t    proximity_domains[MAX_CPUS];
```
No good reason to put the arrays together in this case, right.

debrouxl commented on app/main.c:
```diff
@@ -394,7 +446,7 @@ static void test_all_windows(int my_cpu)
             if (parallel_test) {
                 num_active_cpus = num_enabled_cpus;
-                if(display_mode == DISPLAY_MODE_NA) {
+                if (display_mode == DISPLAY_MODE_NA) {
```
Ah. On other code bases, I prefer doing trivial coding style or comment fixes as drive-by fixes, so as to minimize commit churn.

martinwhitaker replied:
I used to be of the same opinion - mainly because I know I'll forget them if I don't fix them there and then - but it makes reviewing harder.

debrouxl commented on system/smp.c:
```diff
@@ -427,6 +503,187 @@ static bool find_cpus_in_madt(void)
+        if (entry_header->type == SRAT_PROCESSOR_APIC_AFFINITY) {
+            srat_processor_lapic_affinity_entry_t *entry = (srat_processor_lapic_affinity_entry_t *)tab_entry_ptr;
+            if (entry->flags & SRAT_PAAF_ENABLED) {
+                int found1;
```
You're right, that ought to be possible, and it is on GCC 11-13. However, GCC 10 and older barf, even in forced `-std=c17` mode:
```
../system/smp.c: In function ‘find_numa_nodes_in_srat’:
../system/smp.c:608:17: error: a label can only be part of a statement and a declaration is not a statement
  608 |                 int found1 = -1;
      |                 ^~~
```
With the GCC 9 in Ubuntu 20.04 LTS, that would have been a CI failure.

martinwhitaker replied:
Annoying, but I guess it has to stay the way it is.

debrouxl commented on system/smp.c:
```diff
@@ -427,6 +503,187 @@ static bool find_cpus_in_madt(void)
+            if (entry->flags & SRAT_PXAAF_ENABLED) {
+                proximity_domain = entry->proximity_domain;
+                apic_id = entry->apic_id;
+                goto find_proximity_domain;
```
I suppose I'll do that if you really want me to, but I didn't do it at first for two reasons:
* when reading code, I have an easier time following control flow which stays in the same context, on the same screen page, than control flow to an out of line function defined at another place;
* out of line calls to simple functions are typically the kind of low-hanging optimization fruit I've already performed on various places of this code base, so I'm trying not to add them :)
That `smp_narrow_to_proximity_domain()` is out of line despite the fact that it has a single caller, because of a sufficient number of lines, different Translation Units, private variables and no LTO (couldn't make that work), is bad enough :(

martinwhitaker replied:
I won't insist. Early indoctrination makes me avoid gotos as much as possible, but sometimes they are the best solution.

debrouxl commented:
I've just pushed a version which fixes most of Martin's review comments. I'll resolve these afterwards.
Diff between 72353e5 and 7a46d87 (whitespace possibly mangled):
```
diff --git a/app/main.c b/app/main.c
index d7c18be..72b8165 100644
--- a/app/main.c
+++ b/app/main.c
@@ -102,7 +102,6 @@ static int              test_stage = 0;
 // These are exposed in test.h.
 
 uint8_t     chunk_index[MAX_CPUS];
-uint8_t     chunk_index_in_proximity_domain[MAX_CPUS];
 
 int         num_active_cpus = 0;
 int         num_enabled_cpus = 1;
@@ -114,8 +113,8 @@ barrier_t   *run_barrier = NULL;
 spinlock_t  *error_mutex = NULL;
 
 vm_map_t    vm_map[MAX_MEM_SEGMENTS];
-uint32_t    proximity_domains[MAX_CPUS];
 int         vm_map_size = 0;
+uint32_t    proximity_domains[MAX_CPUS];
 
 int         pass_num = 0;
 int         test_num = 0;
@@ -244,6 +243,11 @@ static void global_init(void)
 
     smp_init(smp_enabled);
 
+    // Force disable the NUMA code paths when no proximity domain was found.
+    if (num_proximity_domains == 0) {
+        enable_numa = false;
+    }
+
     // At this point we have started reserving physical pages in the memory
     // map for data structures that need to be permanently pinned in place.
     // This may overwrite any data structures passed to us by the BIOS and/or
@@ -271,9 +275,8 @@ static void global_init(void)
         if (cpu_state[i] == CPU_STATE_ENABLED) {
             if (enable_numa) {
                 uint32_t proximity_domain_idx = smp_get_proximity_domain_idx(i);
-                chunk_index_in_proximity_domain[i] = smp_alloc_cpu_in_proximity_domain(proximity_domain_idx);
-            }
-            else {
+                chunk_index[i] = smp_alloc_cpu_in_proximity_domain(proximity_domain_idx);
+            } else {
                 chunk_index[i] = num_enabled_cpus;
             }
             num_enabled_cpus++;
@@ -382,21 +385,13 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
         }
         if (seg_start < seg_end && seg_start < win_end && seg_end > win_start) {
             // We need to test part of that physical memory segment.
-            if (!enable_numa) {
-non_numa_vm_map_entry:
-                num_mapped_pages += seg_end - seg_start;
-                vm_map[vm_map_size].pm_base_addr = seg_start;
-                vm_map[vm_map_size].start        = first_word_mapping(seg_start);
-                vm_map[vm_map_size].end          = last_word_mapping(seg_end - 1, sizeof(testword_t));
-                vm_map_size++;
-            }
-            else {
+            if (enable_numa) {
                 // Now also pay attention to proximity domains, which are based on physical addresses.
-                uintptr_t orig_start = seg_start << PAGE_SHIFT;
-                uintptr_t orig_end = seg_end << PAGE_SHIFT;
+                uint64_t orig_start = seg_start << PAGE_SHIFT;
+                uint64_t orig_end = seg_end << PAGE_SHIFT;
                 uint32_t proximity_domain_idx;
-                uintptr_t new_start;
-                uintptr_t new_end;
+                uint64_t new_start;
+                uint64_t new_end;
 
                 while (1) {
                     if (smp_narrow_to_proximity_domain(orig_start, orig_end, &proximity_domain_idx, &new_start, &new_end)) {
@@ -411,23 +406,31 @@ non_numa_vm_map_entry:
                             // Proceed to the next part of the range.
                             orig_start = new_end; // No shift here, we already have a physical address.
                             orig_end = seg_end << PAGE_SHIFT;
-                        }
-                        else {
+                        } else {
                             // We're done with this range.
                             break;
                         }
-                    }
-                    else {
+                    } else {
                         // Could not match with proximity domain, fall back to default behaviour. This shouldn't happen !
+                        vm_map[vm_map_size].proximity_domain_idx = 0;
                         goto non_numa_vm_map_entry;
                     }
                 }
+            } else {
+non_numa_vm_map_entry:
+                num_mapped_pages += seg_end - seg_start;
+                vm_map[vm_map_size].pm_base_addr = seg_start;
+                vm_map[vm_map_size].start        = first_word_mapping(seg_start);
+                vm_map[vm_map_size].end          = last_word_mapping(seg_end - 1, sizeof(testword_t));
+                vm_map_size++;
             }
         }
     }
-    /*for (int i = 0; i < vm_map_size; i++) {
+#if 0
+    for (int i = 0; i < vm_map_size; i++) {
         do_trace(0, "vm %0*x - %0*x", 2*sizeof(uintptr_t), vm_map[i].start, 2*sizeof(uintptr_t), vm_map[i].end);
-    }*/
+    }
+#endif
 }
 
 static void test_all_windows(int my_cpu)
diff --git a/app/test.h b/app/test.h
index 3f8fa84..7267273 100644
--- a/app/test.h
+++ b/app/test.h
@@ -22,15 +22,10 @@
 /**
  * A mapping from a CPU core number to the index number of the memory chunk
  * it operates on when performing a memory test in parallel across all the
- * enabled cores.
+ * enabled cores (in the current proximity domain, when NUMA awareness is
+ * enabled).
  */
 extern uint8_t chunk_index[MAX_CPUS];
-/**
- * A mapping from a CPU core number to the index number of the memory chunk
- * it operates on when performing a memory test in parallel across all the
- * enabled cores in the current proximity domain.
- */
-extern uint8_t chunk_index_in_proximity_domain[MAX_CPUS];
 /**
  * An array where the count of used CPUs in the current proximity domain.
  */
diff --git a/boot/boot.h b/boot/boot.h
index 9a65aef..15ca827 100644
--- a/boot/boot.h
+++ b/boot/boot.h
@@ -89,6 +89,8 @@ extern uint8_t        _stacks[];
 
 extern uint8_t _end[];
 
+#define ARRAY_SIZE(var_) (sizeof(var_) / sizeof(var_[0]))
+
 #endif /* ! __ASSEMBLY__ */
 
 #endif /* BOOT_H */
diff --git a/system/smp.c b/system/smp.c
index 6b74ef2..c7662be 100644
--- a/system/smp.c
+++ b/system/smp.c
@@ -542,8 +542,8 @@ static bool find_numa_nodes_in_srat(void)
             srat_memory_affinity_entry_t *entry = (srat_memory_affinity_entry_t *)tab_entry_ptr;
             if (entry->flags & SRAT_MAF_ENABLED) {
                 uint32_t proximity_domain = entry->proximity_domain;
-                uintptr_t start = entry->base_address;
-                uintptr_t end = entry->base_address + entry->address_length;
+                uint64_t start = entry->base_address;
+                uint64_t end = entry->base_address + entry->address_length;
                 int found = -1;
 
                 if (start > end) {
@@ -560,25 +560,23 @@ static bool find_numa_nodes_in_srat(void)
                 }
                 if (found == -1) {
                     // Not found, allocate entry.
-                    if (num_proximity_domains < (int)(sizeof(proximity_domains) / sizeof(proximity_domains[0]))) {
+                    if (num_proximity_domains < (int)(ARRAY_SIZE(proximity_domains))) {
                         proximity_domains[num_proximity_domains] = proximity_domain;
                         found = num_proximity_domains;
                         num_proximity_domains++;
-                    }
-                    else {
+                    } else {
                         // TODO Display message ?
                         return false;
                     }
                 }
 
                 // Now that we have the index of the entry in proximity_domains in found, use it.
-                if (num_memory_affinity_ranges < (int)(sizeof(memory_affinity_ranges) / sizeof(memory_affinity_ranges[0]))) {
+                if (num_memory_affinity_ranges < (int)(ARRAY_SIZE(memory_affinity_ranges))) {
                     memory_affinity_ranges[num_memory_affinity_ranges].proximity_domain_idx = (uint32_t)found;
                     memory_affinity_ranges[num_memory_affinity_ranges].start = start;
                     memory_affinity_ranges[num_memory_affinity_ranges].end = end;
                     num_memory_affinity_ranges++;
-                }
-                else {
+                } else {
                     // TODO Display message ?
                     return false;
                 }
@@ -588,8 +586,7 @@ static bool find_numa_nodes_in_srat(void)
             if (entry_header->length != sizeof(srat_processor_lx2apic_affinity_entry_t)) {
                 return false;
             }
-        }
-        else {
+        } else {
             return false;
         }
         tab_entry_ptr += entry_header->length;
@@ -778,28 +775,28 @@ static bool start_cpu(int cpu_num)
 
 void smp_init(bool smp_enable)
 {
-    for (size_t i = 0; i < sizeof(apic_id_to_cpu_num) / sizeof(apic_id_to_cpu_num[0]); i++) {
+    for (int i = 0; i < (int)(ARRAY_SIZE(apic_id_to_cpu_num)); i++) {
         apic_id_to_cpu_num[i] = 0;
     }
-    for (size_t i = 0; i < sizeof(apic_id_to_proximity_domain_idx) / sizeof(apic_id_to_proximity_domain_idx[0]); i++) {
+    for (int i = 0; i < (int)(ARRAY_SIZE(apic_id_to_proximity_domain_idx)); i++) {
         apic_id_to_proximity_domain_idx[i] = 0;
     }
 
-    for (size_t i = 0; i < sizeof(cpu_num_to_apic_id) / sizeof(cpu_num_to_apic_id[0]); i++) {
+    for (int i = 0; i < (int)(ARRAY_SIZE(cpu_num_to_apic_id)); i++) {
         cpu_num_to_apic_id[i] = 0;
     }
 
-    for (size_t i = 0; i < sizeof(memory_affinity_ranges) / sizeof(memory_affinity_ranges[0]); i++) {
+    for (int i = 0; i < (int)(ARRAY_SIZE(memory_affinity_ranges)); i++) {
         memory_affinity_ranges[i].proximity_domain_idx = UINT32_C(0xFFFFFFFF);
         memory_affinity_ranges[i].start = 0;
         memory_affinity_ranges[i].end = 0;
     }
 
-    for (size_t i = 0; i < sizeof(cpus_in_proximity_domain) / sizeof(cpus_in_proximity_domain[0]); i++) {
+    for (int i = 0; i < (int)(ARRAY_SIZE(cpus_in_proximity_domain)); i++) {
         cpus_in_proximity_domain[i] = 0;
     }
 
-    for (size_t i = 0; i < sizeof(used_cpus_in_proximity_domain) / sizeof(used_cpus_in_proximity_domain[0]); i++) {
+    for (int i = 0; i < (int)(ARRAY_SIZE(used_cpus_in_proximity_domain)); i++) {
         used_cpus_in_proximity_domain[i] = 0;
     }
 
@@ -841,18 +838,6 @@ void smp_init(bool smp_enable)
         cpus_in_proximity_domain[proximity_domain_idx]++;
     }
 
-    // Create dummy entries when there's no (proper) memory affinity information.
-    if (num_memory_affinity_ranges == 0 || num_proximity_domains == 0) {
-        extern uintptr_t pm_limit_lower;
-        extern uintptr_t pm_limit_upper;
-        proximity_domains[0] = 0;
-        num_proximity_domains = 1;
-        memory_affinity_ranges[0].proximity_domain_idx = 0;
-        memory_affinity_ranges[0].start = pm_limit_lower << PAGE_SHIFT;
-        memory_affinity_ranges[0].end = pm_limit_upper << PAGE_SHIFT;
-        num_memory_affinity_ranges = 1;
-    }
-
     // Allocate a page of low memory for AP trampoline and sync objects.
     // These need to remain pinned in place during relocation.
     smp_heap_page = heap_alloc(HEAP_TYPE_LM_1, PAGE_SIZE, PAGE_SIZE) >> PAGE_SHIFT;
@@ -926,7 +911,7 @@ uint32_t smp_get_proximity_domain_idx(int cpu_num)
     return num_available_cpus > 1 ? apic_id_to_proximity_domain_idx[cpu_num_to_apic_id[cpu_num]] : 0;
 }
 
-int smp_narrow_to_proximity_domain(uintptr_t start, uintptr_t end, uint32_t * proximity_domain_idx, uintptr_t * new_start, uintptr_t * new_end)
+int smp_narrow_to_proximity_domain(uint64_t start, uint64_t end, uint32_t * proximity_domain_idx, uint64_t * new_start, uint64_t * new_end)
 {
     for (int i = 0; i < num_memory_affinity_ranges; i++) {
         uint64_t range_start = memory_affinity_ranges[i].start;
@@ -941,8 +926,7 @@ int smp_narrow_to_proximity_domain(uintptr_t start, uintptr_t end, uint32_t * pr
                     *new_start = start;
                     *new_end = end;
                     return 1;
-                }
-                else {
+                } else {
                     // range_start start range_end end.
                     // The given vm_map range needs to be shortened.
                     *proximity_domain_idx = memory_affinity_ranges[i].proximity_domain_idx;
@@ -950,26 +934,22 @@ int smp_narrow_to_proximity_domain(uintptr_t start, uintptr_t end, uint32_t * pr
                     *new_end = range_end;
                     return 1;
                 }
-            }
-            else {
+            } else {
                 // range_start range_end start end
                 // Do nothing, skip to next memory affinity range.
             }
-        }
-        else {
+        } else {
             if (end < range_start) {
                 // start end range_start range_end.
                 // Do nothing, skip to next memory affinity range.
-            }
-            else {
+            } else {
                 if (end <= range_end) {
                     // start range_start end range_end.
                     *proximity_domain_idx = memory_affinity_ranges[i].proximity_domain_idx;
                     *new_start = start;
                     *new_end = range_start;
                     return 1;
-                }
-                else {
+                } else {
                     // start range_start range_end end.
                     *proximity_domain_idx = memory_affinity_ranges[i].proximity_domain_idx;
                     *new_start = start;
@@ -983,12 +963,14 @@ int smp_narrow_to_proximity_domain(uintptr_t start, uintptr_t end, uint32_t * pr
     return 0;
 }
 
-/*void get_memory_affinity_entry(int idx, uint32_t * proximity_domain_idx, uint64_t * start, uint64_t * end)
+#if 0
+void get_memory_affinity_entry(int idx, uint32_t * proximity_domain_idx, uint64_t * start, uint64_t * end)
 {
     *proximity_domain_idx = memory_affinity_ranges[idx].proximity_domain_idx;
     *start = memory_affinity_ranges[idx].start;
     *end = memory_affinity_ranges[idx].end;
-}*/
+}
+#endif
 
 barrier_t *smp_alloc_barrier(int num_threads)
 {
diff --git a/system/smp.h b/system/smp.h
index d22f9b9..86b319d 100644
--- a/system/smp.h
+++ b/system/smp.h
@@ -97,7 +97,7 @@ static inline uint8_t smp_alloc_cpu_in_proximity_domain(uint32_t proximity_domai
 /**
  * Computes the first span, limited to a single proximity domain, of the given memory range.
  */
-int smp_narrow_to_proximity_domain(uintptr_t start, uintptr_t end, uint32_t * proximity_domain_idx, uintptr_t * new_start, uintptr_t * new_end);
+int smp_narrow_to_proximity_domain(uint64_t start, uint64_t end, uint32_t * proximity_domain_idx, uint64_t * new_start, uint64_t * new_end);
 
 //int count_cpus_for_proximity_domain_corresponding_to_range(uintptr_t start, uintptr_t end, uint32_t proximity_domain_idx);
 
diff --git a/tests/test_helper.c b/tests/test_helper.c
index ad2b5a2..b9987c5 100644
--- a/tests/test_helper.c
+++ b/tests/test_helper.c
@@ -49,20 +49,18 @@ void calculate_chunk(testword_t **start, testword_t **end, int my_cpu, int segme
                 uintptr_t chunk_size   = round_down(segment_size / used_cpus_in_proximity_domain[proximity_domain_idx], chunk_align);
 
                 // Calculate chunk boundaries.
-                *start = (testword_t *)((uintptr_t)vm_map[segment].start + chunk_size * chunk_index_in_proximity_domain[my_cpu]);
+                *start = (testword_t *)((uintptr_t)vm_map[segment].start + chunk_size * chunk_index[my_cpu]);
                 *end   = (testword_t *)((uintptr_t)(*start) + chunk_size) - 1;
 
                 if (*end > vm_map[segment].end) {
                     *end = vm_map[segment].end;
                 }
-            }
-            else {
+            } else {
                 // Nope.
                 *start = (testword_t *)1;
                 *end = (testword_t *)0;
             }
-        }
-        else {
+        } else {
             uintptr_t segment_size = (vm_map[segment].end - vm_map[segment].start + 1) * sizeof(testword_t);
             uintptr_t chunk_size   = round_down(segment_size / num_active_cpus, chunk_align);
 
```

debrouxl commented on app/main.c:
```diff
@@ -369,13 +381,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
+                    }
+                    else {
+                        // Could not match with proximity domain, fall back to default behaviour. This shouldn't happen !
+                        goto non_numa_vm_map_entry;
```
Branches swapped; usage of uninitialized value fixed.

martinwhitaker commented on app/main.c:
```diff
@@ -369,13 +384,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
+                        if (new_start != orig_start || new_end != orig_end) {
+                            // Proceed to the next part of the range.
+                            orig_start = new_end; // No shift here, we already have a physical address.
+                            orig_end = seg_end << PAGE_SHIFT;
```
And here.

martinwhitaker reviewed (changes requested):
Sorry for the delay in getting back to this. Just one problem with the changes.

debrouxl commented on app/main.c:
```diff
@@ -369,13 +384,53 @@ static void setup_vm_map(uintptr_t win_start, uintptr_t win_end)
+            if (enable_numa) {
+                // Now also pay attention to proximity domains, which are based on physical addresses.
+                uint64_t orig_start = seg_start << PAGE_SHIFT;
+                uint64_t orig_end = seg_end << PAGE_SHIFT;
```
Hmpf. It's not the first time I'm mixing types or forgetting a cast before shifts while working on this code base...
Sadly, GCC's conversion-related and shift-related warnings - `-Wconversion -Wfloat-conversion -Wsign-conversion -Warith-conversion -Wshift-overflow=2` - report *lots* of things, some of which are probably interesting, but not this one...

x86fr commented:
@debrouxl looks ok to me (even if many potential interactions are still possible). Can you please post here your latest benchmarks results to keep track?

debrouxl commented:
OK. I'll redo a couple benchmarks on 4S Bulldozer RS904A, 4S IVB R820 and 2S BDX DL380 Gen9 after rebasing the changes, and regroup the results. Not that the timings should be different from the temporal mode timings I posted in #12 .

debrouxl commented:
Little change since the previous post in #12 , indeed. NUMA-aware mode remains ~13% faster on my DL380 Gen9:
```
HPE DL380 Gen9, 2 x Xeon E5-2667v4 @ 3.2 GHz, 8 x 16 GB DDR4-PC19200R = 128 GB (~ 2 GB memory hole), 16 thread mode (SMT disabled, x2APIC disabled).
2 NUMA nodes of 64 GB (1 per socket), 2 QPI links per socket (pair topology), SLIT contains 10 / 21 values.
memtest86+ 4988b5b (HEAD of numa branch as of 2024/03/04)
Pass 1:
                                 (normal)     numa                 
test  0 (test_addr_walk1):       00:00:00     00:00:00             
test  1 (test_own_addr1):        00:01:06     00:01:06 (= 00:00:00)
test  2 (test_own_addr2):        00:10:01     00:10:02 (+ 00:00:01)
test  3 (test_mov_inv_fixed):    00:18:49     00:18:49 (= 00:00:00)
test  4 (test_mov_inv_fixed):    00:19:54     00:19:43 (- 00:00:11)
test  5 (test_mov_inv_fixed):    00:25:01     00:23:44 (- 00:01:17)
test  6 (test_mov_inv_walk1):    00:30:27     00:28:09 (- 00:02:18)
test  7 (test_block_move):       01:13:56     01:04:28 (- 00:09:28)
test  8 (test_mov_inv_random):   01:18:05     01:07:56 (- 00:10:09)
test  9 (test_modulo_n):         01:23:45     01:12:53 (- 00:10:52)
test 10 (test_bit_fade):         01:50:32     01:36:03 (- 00:14:29)
test  0:                         01:54:18     01:39:48 (- 00:14:30)
```

I've started a normal run on the R820... should take over 2 days, I just hope that there won't be a power outage in the meantime :)

debrouxl commented:
Nearly 56h for checking 256 GB of RAM without NUMA awareness on the R820, as extrapolated from the time taken to check a quarter of the RAM...
I'll edit the message with information from the NUMA-aware run, expected to run in less than 8h again:
```
Dell Poweredge R820, 4 x Xeon E5-4627v2 @ 3.3 GHz, 32 x 8 GB DDR3-PC14900R = 256 GB (2.05 GB memory hole), 32 thread mode (no SMT).
4 NUMA nodes of 64 GB (1 per socket), 2 QPI links per socket (ring topology), SLIT contains 10 / 20 / 30 values.
memtest86+ 4988b5b (HEAD of numa branch as of 2024/03/04)
Pass 1:
                                 (normal)     numa                 
test  0 (test_addr_walk1):       00:00:00     00:00:00             
test  1 (test_own_addr1):        00:08:30     00:  :   (= 00:  :  )
test  2 (test_own_addr2):        01:10:27     0 :  :   (+ 0 :  :  )
test  3 (test_mov_inv_fixed):    02:12:41     0 :  :   (= 0 :  :  )
test  4 (test_mov_inv_fixed):    02:58:41     0 :  :   (- 0 :  :  )
test  5 (test_mov_inv_fixed):    06:48:41     0 :  :   (- 0 :  :  )
test  6 (test_mov_inv_walk1):    10:36:57     0 :  :   (- 0 :  :  )
test  7 (test_block_move):       38:08:35     0 :  :   (- 0 :  :  )
test  8 (test_mov_inv_random):   39:34:40     0 :  :   (- 0 :  :  )
test  9 (test_modulo_n):         43:06:28     0 :  :   (- 0 :  :  )
test 10 (test_bit_fade):         55:51:54     0 :  :   (- 0 :  :  )
test  0:                         55:58:24     0 :  :   (- 0 :  :  )
```
